File: tck/src/main/java/com/akkaserverless/javasdk/tck/model/valueentity/ValueEntityTckModelEntity.java
/*
 * Copyright 2021 Lightbend Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.akkaserverless.javasdk.tck.model.valueentity;

import com.akkaserverless.javasdk.Context;
import com.akkaserverless.javasdk.ServiceCall;
import com.akkaserverless.javasdk.ServiceCallRef;
import com.akkaserverless.javasdk.SideEffect;
import com.akkaserverless.javasdk.valueentity.ValueEntity;
import com.akkaserverless.tck.model.ValueEntity.Persisted;
import com.akkaserverless.tck.model.ValueEntity.Request;
import com.akkaserverless.tck.model.ValueEntity.RequestAction;
import com.akkaserverless.tck.model.ValueEntity.Response;

import java.util.ArrayList;
import java.util.List;

public class ValueEntityTckModelEntity extends ValueEntity<Persisted> {

  private final ServiceCallRef<Request> serviceTwoCall;

  public ValueEntityTckModelEntity(Context context) {
    serviceTwoCall =
        context
            .serviceCallFactory()
            .lookup("akkaserverless.tck.model.valueentity.ValueEntityTwo", "Call", Request.class);
  }

  public Effect<Response> process(Persisted state, Request request) {
    // FIXME the effect API doesn't support all combinations, and that might be fine?
    String value = state.getValue();
    Effect.OnSuccessBuilder<Persisted> builder = null;
    Effect<Response> result = null;
    List<SideEffect> e = new ArrayList<>();
    for (RequestAction action : request.getActionsList()) {
      switch (action.getActionCase()) {
        case UPDATE:
          value = action.getUpdate().getValue();
          builder = effects().updateState(Persisted.newBuilder().setValue(value).build());
          break;
        case DELETE:
          value = "";
          builder = effects().deleteState();
          break;
        case FORWARD:
          if (builder == null) {
            result = effects().forward(serviceTwoRequest(action.getForward().getId()));
          } else {
            result = builder.thenForward(serviceTwoRequest(action.getForward().getId()));
          }
          break;
        case EFFECT:
          com.akkaserverless.tck.model.ValueEntity.Effect effect = action.getEffect();
          e.add(SideEffect.of(serviceTwoRequest(effect.getId()), effect.getSynchronous()));
          break;
        case FAIL:
          result = effects().error(action.getFail().getMessage());
          break;
      }
    }
    if (builder == null && result == null) {
      return effects().reply(Response.newBuilder().setMessage(value).build()).addSideEffects(e);
    }
    if (result == null) {
      return builder.thenReply(Response.newBuilder().setMessage(value).build()).addSideEffects(e);
    } else {
      return result.addSideEffects(e);
    }
  }

  private ServiceCall serviceTwoRequest(String id) {
    return serviceTwoCall.createCall(Request.newBuilder().setId(id).build());
  }

  @Override
  public Persisted emptyState() {
    return Persisted.getDefaultInstance();
  }
}

File: tck/scala-tck/src/main/scala/kalix/tck/model/valueentity/ValueEntityTckModelEntity.scala
/*
 * Copyright 2024 Lightbend Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kalix.tck.model.valueentity

import kalix.scalasdk.SideEffect
import kalix.scalasdk.valueentity.{ ValueEntity, ValueEntityContext }

/** A value entity. */
class ValueEntityTckModelEntity(context: ValueEntityContext) extends AbstractValueEntityTckModelEntity {
  override def emptyState: Persisted = Persisted.defaultInstance

  override def process(currentState: Persisted, request: Request): ValueEntity.Effect[Response] = {
    case class HandlingState(
        state: String,
        effect: Either[ValueEntity.Effect.OnSuccessBuilder[Persisted], ValueEntity.Effect[Response]],
        sideEffect: Seq[SideEffect]) {
      def handle(action: RequestAction): HandlingState =
        action.action match {
          case RequestAction.Action.Empty => this
          case RequestAction.Action.Update(Update(newValue, _)) =>
            copy(state = newValue, effect = Left(effects.updateState(Persisted(newValue))))
          case RequestAction.Action.Delete(Delete(_)) =>
            copy(state = "", effect = Left(effects.deleteEntity()))
          case RequestAction.Action.Fail(Fail(message, _)) =>
            copy(effect = Right(effects.error(message)))
          case RequestAction.Action.Effect(Effect(id, sync, _)) =>
            copy(sideEffect = sideEffect :+ SideEffect(components.valueEntityTwoEntity.call(Request(id)), sync))
          case RequestAction.Action.Forward(Forward(id, _)) =>
            val call = components.valueEntityTwoEntity.call(Request(id))
            val newEffect: Either[ValueEntity.Effect.OnSuccessBuilder[Persisted], ValueEntity.Effect[Response]] =
              effect match {
                case Left(e)  => Right(e.thenForward(call))
                case Right(_) => Right(effects.forward(call))

              }
            copy(effect = newEffect)
        }
      def result: ValueEntity.Effect[Response] =
        (effect match {
          case Left(e)  => e.thenReply(Response(state))
          case Right(e) => e
        }).addSideEffects(sideEffect)
    }

    request.actions
      .foldLeft(HandlingState(currentState.value, Right(effects.reply(Response(currentState.value))), Vector.empty))(
        _.handle(_))
      .result
  }
}
